pkg/sources/gcs/gcs.go

```diff
@@ -141,21 +324,57 @@ func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) err
 			continue
 		}
 
+		if err := s.startProcessing(ctx, progress, o); err != nil {
+			ctx.Logger().V(1).Info("GCS source error starting to process objects", "error", err)
+			continue
+		}
 		wg.Add(1)
-		go func(obj object) {
+		go func(obj object, resume *progressInfo) {
 			defer wg.Done()
 
 			if err := s.processObject(ctx, o); err != nil {
-				ctx.Logger().V(1).Info("GCS source error processing objects", "name", s.name, "error", err)
+				ctx.Logger().V(1).Info("error setting start progress progress", "name", o.name, "error", err)
+				return
+			}
+			if err := s.endProcessing(ctx, resume, o); err != nil {
+				ctx.Logger().V(1).Info("error setting end progress progress", "name", o.name, "error", err)
 			}
-		}(o)
+		}(o, progress)
 	}
 	wg.Wait()
 
 	ctx.Logger().Info("GCS source finished processing", "name", s.name)
 	return nil
 }
 
+func (s *Source) startProcessing(ctx context.Context, progress *progressInfo, o object) error {
+	progress.mu.Lock()
+	defer progress.mu.Unlock()
```

Comment:
**Nit:** It feels a bit odd to reach into the struct to lock it from outside its methods.. if we're locking at this level I think we could add the mutex to `Source` instead of `progressInfo` and add a note to `progressInfo` that it's not thread safe.